src/backend/distributed/utils/resource_lock.c

```diff
@@ -1196,3 +1199,147 @@ AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
+ *
+ * Only tables, views, and foreign tables can be locked with this function. Other relations
+ * will cause an error.
+ *
+ * Skips non distributed relations.
+ * configs parameter is used to configure what relation will be locked and if the lock
+ * should throw an error if it cannot be acquired immediately
+ */
+void
+AcquireDistributedLockOnRelations(List *relationList, LOCKMODE lockMode, uint32 configs)
+{
+	if (!ClusterHasKnownMetadataWorkers() || !EnableMetadataSync || !EnableDDLPropagation)
+	{
+		return;
+	}
+
+	List *distributedRelationList = NIL;
+
+	/*
+	 * this is used to enforce the lock order:
+	 * [...A], [...B]
+	 * Where:
+	 * A: Excplicitly locked relations
+	 * B: Relations which depend on A, but are not locked by the lock
+	 *    command by default
+	 *
+	 * e.g.
+	 *  Let view v with tables t1, t2
+	 *  for LOCK v IN * MODE;
+	 *  A: v
+	 *  B: nil
+	 *  the lock command will lock t1 and t2 so no need to populate B
+	 *
+	 *  Let tables t1 and t2 with t2 referencing t1 with a foreign key
+	 *  for TRUNCATE t1 (we need to acquire locks on t1 and t2)
+	 *  A: t1
+	 *  B: t2
+	 *  t2 will not be locked by lock command if we call LOCK t1
+	 *  so we need to add it to B
+	 */
+	List *dependentRelationIds = NIL;
+
+	RangeVar *rangeVar = NULL;
+	foreach_ptr(rangeVar, relationList)
+	{
+		Oid relationId = RangeVarGetRelid(rangeVar, NoLock, false);
+
+		/*
+		 * Note that allowing the user to lock shards could lead to
+		 * distributed deadlocks due to shards not being locked when
+		 * a distributed table is locked.
+		 * However, because citus.enable_manual_changes_to_shards
+		 * is a guc which is not visible by default, whoever is using this
+		 * guc will hopefully know what they're doing and avoid such scenarios.
+		 */
+		ErrorIfIllegallyChangingKnownShard(relationId);
+
+		/*
+		 * we want to prevent under privileged users to trigger establishing connections,
+		 * that's why we have this additional check. Otherwise, we'd get the errors as
+		 * soon as we execute the command on any node
+		 */
+		EnsureCanAcquireLock(relationId, lockMode);
+
+		if (get_rel_relkind(relationId) == RELKIND_VIEW)
+		{
+			ObjectAddress viewAddress = { 0 };
+			ObjectAddressSet(viewAddress, RelationRelationId, relationId);
+
+			if (!IsObjectDistributed(&viewAddress))
+			{
+				continue;
+			}
+		}
+		else if (!ShouldSyncTableMetadata(relationId))
+		{
+			continue;
+		}
+
+		distributedRelationList = list_append_unique_oid(distributedRelationList,
+														 relationId);
+
+		if ((configs & DIST_LOCK_REFERENCING_TABLES) > 0)
+		{
+			CitusTableCacheEntry *cacheEntry = GetCitusTableCacheEntry(relationId);
+			Assert(cacheEntry != NULL);
+
+			List *referencingTableList = cacheEntry->referencingRelationsViaForeignKey;
+
+			Oid referencingRelationId = InvalidOid;
+			foreach_oid(referencingRelationId, referencingTableList)
+			{
+				dependentRelationIds = lappend_oid(dependentRelationIds,
```

Comment:
what if `referencingRelationId` returns false for `ShouldSyncTableMetadata`? Maybe we can add that check as well

Reply:
My reasoning was that citus tables need to have other citus tables as dependencies, but this check does not hurt to have us covered for potential future uses of the function.